menelaus/injection/class_manipulation.py

```diff
@@ -94,3 +96,119 @@ def class_join(data, target_col, class_1, class_2, new_class, from_index, to_ind
         class_idx = class_idx[(class_idx < to_index) & (class_idx >= from_index)]
         ret[class_idx, target_col] = new_class
     return ret
+
+# endregion
+
+
+# region - LTF-inspired class manipulation functions
+
+def class_probability_shift(data, target_col, from_index, to_index, classes, p_classes):
+    """
+    Resamples the data over a specified window, with altered probability
+    for specified classes (and uniform probability for remaining classes).
+    Accepts ``pandas.DataFrame`` with column names or ``numpy.ndarray``
+    with column indices.
+
+    Ref. :cite:t:`LTFmethods`
+
+    Args:
+        data (np.array): data to inject with drift
+        target_col (int or str): column index/label of targets column
+        from_index: row index at which to start class swap
+        to_index: row index at which to end (non-inclusive) class swap
+        classes (list[int]): list of classes with altered sampling chance
+        p_classes (float): altered sampling chance for members of ``classes``
+
+    Returns:
+        np.array or pd.DataFrame: copy of data, resampled with shifted
+            class probability for 1 or more desired classes
+    """
+    ret = np.copy(data)
+    classes = np.unique(ret[:, target_col])
+
+    # distribution for each data point, and reordering of each point by class
+    p_distribution = np.array([])
+    sample_grouped = np.array([])
+
+    # locate all classes to shift
+    shifted_classes_idx = np.where(data[:, target_col].isin(classes))[0]
+
+    # locate each class in window
+    for cls in classes:
+        cls_idx = np.where(data[:, target_col] == cls)[0]
+        cls_idx = cls_idx[
+            (cls_idx < to_index) &
+            (cls_idx >= from_index)
+        ]
+
+        if cls in classes:
+            # pts from shifted classes drawn with separate probability
+            p_individual = p_classes / shifted_classes_idx.shape[0]
+        else:
+            # pts from remaining classes drawn with uniform probability
+            p_individual = ((1 - p_classes) / (classes.shape[0] - 1)) / cls_idx.shape[0]
+
+        # append to grouped array and corresponding distribution
+        sample_grouped = np.concatenate((sample_grouped, data[cls_idx]))
+        p_distribution = np.concatenate(p_distribution, np.ones(cls_idx.shape[0]) * p_individual)
+
+    # shuffled sample over window, with replacement, with weights
+    sample = np.random.choice(sample_grouped, to_index-from_index, True, p_distribution)
+    ret[from_index:to_index] = sample
+    return ret
+
+
+def class_dirichlet_shift(data, target_col, from_index, to_index, alpha):
```

Comment:
and to the best of my understanding of dirichlet, this one seems implemented correctly too. 
To double check the way I'm understanding it.. 
The general functionality is to change labels so they appear in proportions as determined by a dirichlet distribution. In the input dictionary, users specify parameters for dirchlet for each class.
- Like previous function, we may need to check that the inputted proportions in alpha exactly equals 1

Reply:
Yep, that's the reasoning I used and what I'm thinking the `numpy` function is doing.

Reply:
@indialindsay Since the alpha interpretation I get from `numpy` is an integer weights sort of thing, does this add-to-1 check still apply? I was thinking something like

```
1 + 10 + 2 = 13
proportions are 1, 10, 2 out of 13
```

In which case a check wouldn't make sense -- do you have a different interpretation?

Reply:
yes you're right, a check doesn't make sense when using the dirichlet function.. think it's all good to not add that in